File: answers/python/exercise1/SSLClient.py
import socket
import ssl
from pathlib import Path


LOCAL_HOST = 'localhost'
LOCAL_PORT = 8282
RESOURCE_DIRECTORY = Path(__file__).resolve().parent.parent / 'resources' / 'client'
CA_CERT = RESOURCE_DIRECTORY / 'ca.cert.pem'


def main():
    """
    Exercise 1 Client
    """
    # Create a standard TCP Socket
    sock = socket.socket(socket.AF_INET)

    # Create SSL context which holds the parameters for any sessions
    context = ssl.create_default_context(ssl.Purpose.SERVER_AUTH)
    context.check_hostname = False
    context.load_verify_locations(CA_CERT)
    
    # Restrict available protocols and ciphers
    context.options |= ssl.OP_NO_TLSv1 | ssl.OP_NO_TLSv1_1 | ssl.OP_NO_SSLv3
    context.set_ciphers('ALL:!DSS:!DHE:!aNULL:!eNull')

    # We can wrap in an SSL context first, then connect
    conn = context.wrap_socket(sock, server_hostname=LOCAL_HOST)
    try:
        # Handshake - conn is an SSLSocket
        conn.connect((LOCAL_HOST, LOCAL_PORT))

        # What parameters were established?
        print("Negotiated session using cipher suite: {0}\n".format(conn.cipher()[0]))

        # In python sockets send and receive bytes. Send some numbers:
        conn.send(bytes([2, 3, 5, 7, 11, 13, 17, 19, 23]))

        # Receive a number back from the server
        server_response = conn.recv(1024)

        # Server response is an int, convert it back
        print(int.from_bytes(server_response, 'big'))
    finally:
        conn.close()


if __name__ == '__main__':
    main()

File: answers/java/exercise1/SSLClient.java
package com.tlsexercises.exercise1;

import java.io.*;
import javax.net.ssl.*;
import java.nio.*;

/**
 * Exercise 1 Client
 */
public class SSLClient {

    private static final String REMOTE_HOST = "localhost";
    private static final int REMOTE_PORT = 8282;
    private static final String TRUSTSTORE_LOCATION = SSLClient.class.getResource("/client/ClientTrustStore.jks").getPath();
    private static final String TRUSTSTORE_PASSWORD = "experttls";

    // Entry point
    public static void main(String[] args) {
        // You can also update properties as a command line parameter e.g.
        //     -Djavax.net.ssl.keyStore="keystore_location"
        System.setProperty("javax.net.ssl.trustStore", TRUSTSTORE_LOCATION);
        System.setProperty("javax.net.ssl.trustStorePassword", TRUSTSTORE_PASSWORD);

        // Obtain the default socket factory
        SSLSocketFactory f = (SSLSocketFactory) SSLSocketFactory.getDefault();

        try {
            // Create a socket - will not connect yet
            SSLSocket socket = (SSLSocket) f.createSocket(REMOTE_HOST, REMOTE_PORT);

            // Set protocols and suites
            socket.setEnabledCipherSuites(new String[]{ "TLS_AES_128_GCM_SHA256", "TLS_CHACHA20_POLY1305_SHA256", "TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256" });
            socket.setEnabledProtocols(new String[]{ "TLSv1.3", "TLSv1.2" });

            // Handshake to create a session
            socket.startHandshake();

            // What parameters were established?
            System.out.println(String.format("Negotiated Session: %s", socket.getSession().getProtocol()));
            System.out.println(String.format("Cipher Suite: %s", socket.getSession().getCipherSuite()));

            // We're reading and writing bytes. Other streams can be used.
            BufferedOutputStream output = new BufferedOutputStream(socket.getOutputStream());
            BufferedInputStream input = new BufferedInputStream(socket.getInputStream());
            output.write(new byte[] { 2, 3, 5, 7, 11, 13, 17, 19, 23 });
            output.flush();

            // Read the server response up to a max of 64 bytes.
            byte[] serverResponse = new byte[64];
            int len = input.read(serverResponse, 0, 64);

            // Server response is an int, so should be 4 bytes
            if (len == 4) {
                ByteBuffer result = ByteBuffer.allocate(4).put(serverResponse, 0, 4);
                System.out.println("\nServer result: " + result.getInt(0));
            }

            input.close();
            output.close();
            socket.close();
        } catch (Exception e) {
            System.err.println("Exception: " + e.toString());
        }
    }
}